File: 36.StockSpanAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio       */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

public class StockSpan {

    // Function to calculate stock span
    public static int[] calculateSpan(int[] prices) {
        int n = prices.length;          // total number of days
        int[] span = new int[n];        // result array for span of each day
        Stack<Integer> st = new Stack<>();  // stack to store indices of days

        // Span of first day is always 1
        span[0] = 1;
        st.push(0);  // push index of first day

        // Traverse the array from day 1 to day n-1
        for (int i = 1; i < n; i++) {
            
            // Pop elements while current price is higher or equal
            while (!st.isEmpty() && prices[i] >= prices[st.peek()]) {
                st.pop();
            }

            // If stack is empty → all previous prices are smaller
            if (st.isEmpty()) {
                span[i] = i + 1;
            } else {
                // Otherwise, span is distance between current index and last higher price index
                span[i] = i - st.peek();
            }

            // Push this index into the stack
            st.push(i);
        }

        return span;
    }

    public static void main(String[] args) throws IOException {
        // Example prices
        int[] prices = {100, 80, 60, 70, 60, 75, 85};

        // Call the function
        int[] result = calculateSpan(prices);

        // Print prices
        System.out.println("Prices: " + Arrays.toString(prices));
        // Print spans
        System.out.println("Span:   " + Arrays.toString(result));
    }
}

File: 36.StockSpanAlgo.py
# The Stock Span Algorithm is used to find the number of consecutive days before the 
# current day for which the stock price was 
# less than or equal to today’s price. It helps analyze stock trends and investor confidence.

# Explanation

# Imagine you have stock prices for n days.

# For each day, we want to know: How many days in a row (including today) the price was not lower than today’s price?

# Naive approach: For each day, look backwards until you find a higher price → O(n²).

# Optimized approach (using stack):

# Use a stack to keep track of indices of days.

# Traverse prices one by one:

# While stack is not empty and current price ≥ price at top of stack → pop.

# If stack becomes empty → span is i+1.

# Otherwise, span is i - top_of_stack.

# Push current index onto the stack.

# This reduces time to O(n).

def stock_span(prices):
    n = len(prices)
    span = [0] * n
    st = []

    span[0] = 1
    st.append(0)

    for i in range(1, n):
        while st and prices[i] >= prices[st[-1]]:
            st.pop()

        if not st:
            span[i] = i + 1
        else:
            span[i] = i - st[-1]

        st.append(i)

    return span

prices = [100, 80, 60, 70, 60, 75, 85]
result = stock_span(prices)
print("Prices:", prices)
print("Span: ", result)

        



